libs/sr-ui/src/Hook/useEquippedRelics.ts

```diff
@@ -1,54 +1,40 @@
-import type { CharacterKey, RelicSlotKey } from '@genshin-optimizer/sr-consts'
-import type { ICachedRelic } from '@genshin-optimizer/sr-db'
-import { useEffect, useState } from 'react'
+import {
+  allRelicSlotKeys,
+  type RelicSlotKey,
+} from '@genshin-optimizer/sr-consts'
+import type {
+  ICachedRelic,
+  ICachedSroCharacter,
+} from '@genshin-optimizer/sr-db'
+import { objKeyMap, objMap } from '@genshin-optimizer/util'
+import { useEffect, useMemo, useState } from 'react'
 import { useDatabaseContext } from '../Context'
-import { useCharacter } from './useCharacter'
 
-export function useEquippedRelics(characterKey: CharacterKey | '') {
+export function useEquippedRelics(
+  equippedRelics: ICachedSroCharacter['equippedRelics'] | undefined
+) {
+  const relicSlotIds = useMemo(
+    () =>
+      equippedRelics ? equippedRelics : objKeyMap(allRelicSlotKeys, () => ''),
+    [equippedRelics]
+  )
   const { database } = useDatabaseContext()
-  const character = useCharacter(characterKey)
   const [relics, setRelics] = useState<
     Record<RelicSlotKey, ICachedRelic | undefined>
-  >({} as Record<RelicSlotKey, ICachedRelic>)
-  useEffect(() => {
-    console.log(
-      character?.equippedRelics &&
-        (Object.fromEntries(
-          Object.entries(character?.equippedRelics).map(
-            ([slotKey, relicId]) => [
-              slotKey,
-              { ...database.relics.get(relicId) },
-            ]
-          )
-        ) as Record<RelicSlotKey, ICachedRelic | undefined>)
-    )
-    return (
-      character?.equippedRelics &&
-      setRelics(
-        Object.fromEntries(
-          Object.entries(character?.equippedRelics).map(
-            ([slotKey, relicId]) => [slotKey, database.relics.get(relicId)]
-          )
-        ) as Record<RelicSlotKey, ICachedRelic | undefined>
-      )
-    )
-  }, [database, characterKey, character?.equippedRelics])
+  >(() => objMap(relicSlotIds, (id) => database.relics.get(id)))
 
   useEffect(() => {
-    const relicIds = character?.equippedRelics
-      ? Object.values(character.equippedRelics)
-      : []
-    if (relicIds) {
-      relicIds.forEach((relicId) =>
-        database.relics.follow(
```

Comment:
you are registering for the event listener with `follow`, but it actually returns a function to unfollow, which you are not using.